### PR TITLE
Task 3: Show books

### PR DESCRIPTION
**Pull request by rka-wtag**
**Description**

- Added Show books component for Showing books in the book section
- Added service for books to make API calls

image

### CODE REVIEW

File: src/app/Home/book/show-books/show-books.component.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    this.bookSevice.fetchBooks().subscribe({
+      next: (res) => {
+        res.forEach(book => {
+          this.books.push(book);
+          this.genresMap.set(book.id, book.genres.map(genre => genre.name).join(', '));
+        })
+      },
+      error: (err) => {
+        //error from backend
+        //currently facing problem getting errors from backend
```
Comment (rka-wtag) [inline src/app/Home/book/show-books/show-books.component.ts:40]: Here, Having problems gettiing errors from backend. Currently the value of `err` it is receiving is "OK" string which was never sent from backend.

File: src/app/Home/book/show-books/show-books.component.css [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    flex-wrap: wrap;
+}
+
+.book-row {
+    display: flex;
+    align-items: flex-start;
+    margin: 10px;
+    padding: 15px;
+    border: 1px solid #ccc;
+    border-radius: 5px;
```
Comment (ich-wtag) [inline src/app/Home/book/show-books/show-books.component.css]:
```suggestion
    margin: 0.62rem 0;
    padding: 0.94rem;
```
When we write like **margin: 10px;** (or **padding:0.94rem**) this, it actually creates margin (or padding) to top, left, bottom and right respectively. If you don't need any margin to left and right side we can write like as suggested. If you need it then you can keep as it is now.

**ditto**
please convert all the **px** to **rem** unit.

File: src/app/Home/book/show-books/show-books.component.css [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    width: 100%; /* Ensure each row takes up the full width */
```
Comment (ich-wtag) [inline src/app/Home/book/show-books/show-books.component.css]: Here do we need this? As far as I know **block level element ** by default takes 100% width, if it's parent don't has the padding or specified width. 

another thing, you can add 
*,
*::after,
*::before {
margin:0,
padding:0,
box-sizing: border-box
}

at your **root**, **style.css** file. That may neutralize the default margin, padding and box-sizing problem and may solve this width problem too.

File: src/app/Home/book/show-books/show-books.component.css [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+.book-column {
+    flex: 1; /* Take up available space */
+}
+
+.book-title {
+    font-size: 1.2rem;
+    font-weight: bold;
+    margin-bottom: 10px;
```
Comment (ich-wtag) [inline src/app/Home/book/show-books/show-books.component.css]: ditto

File: src/app/Home/book/show-books/show-books.component.css [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    justify-content: space-between;
+}
+
+.edit-button,
+.delete-button {
+    margin-top: 10px;
+    padding: 5px 8px;
+    cursor: pointer;
+    border: none;
+    border-radius: 5px;
```
Comment (ich-wtag) [inline src/app/Home/book/show-books/show-books.component.css]: ditto

File: src/app/Home/book/show-books/show-books.component.css (hunk 3 above)
Comment (rka-wtag) [reply src/app/Home/book/show-books/show-books.component.css]: bhai this has already been applied to my style.css file

**Review by ich-wtag — APPROVED** [review]

File: src/app/services/Book/book.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { TestBed } from '@angular/core/testing';
+
```
Comment (ban-wtag) [inline src/app/services/Book/book.service.spec.ts]: remove empty line

File: src/app/Home/book/show-books/show-books.component.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+          this.genresMap.set(book.id, book.genres.map(genre => genre.name).join(', '));
+        })
+      },
+      error: (err) => {
+      }
+    })
+  }
+
+
+  editBook(id: number) {
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.ts]: If this method is not being used in this branch, then we can remove it and add it to the respective branch. thank you.

File: src/app/Home/book/show-books/show-books.component.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  deleteBook(id: number) {
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.ts]: [ditto](https://github.com/rka-wtag/lms-frontend/pull/3/files#r1501742463)

File: src/app/Home/book/show-books/show-books.component.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  fetchBooks(): void {
+    this.bookSevice.fetchBooks().subscribe({
+      next: (res) => {
+        res.forEach(book => {
+          this.books.push(book);
+          this.genresMap.set(book.id, book.genres.map(genre => genre.name).join(', '));
+        })
+      },
+      error: (err) => {
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.ts]: If possible, we can display a toast message when any error occurs.

File: src/app/Home/book/show-books/show-books.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.spec.ts]: [ditto](https://github.com/rka-wtag/lms-frontend/pull/3/files#r1501742223)

File: src/app/Home/book/show-books/show-books.component.html [hunk 11]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+    <app-sidebar/>
+    <div class="book-list">
+        <div *ngFor="let book of books" class="book-row">
+            <div class="book-column">
+                <div class="book-title">{{ book.title }}</div>
+                <div class="book-details">
+                    <div><strong>Author:</strong> {{ book.author.firstName }}</div>
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.html]: using css for styling would be better choice.

File: src/app/Home/book/show-books/show-books.component.html [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+                    <div><strong>Genres:</strong> {{ genresMap.get(book.id) }}</div>
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.html]: [ditto](https://github.com/rka-wtag/lms-frontend/pull/3/files#r1501742818)

File: src/app/Home/book/show-books/show-books.component.html [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+                    <div><strong>Publication Year:</strong> {{ book.publicationYear }}</div>
+                    <div><strong>Available Copies:</strong> {{ book.copiesAvailable }}</div>
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.html]: [ditto](https://github.com/rka-wtag/lms-frontend/pull/3/files#r1501742818)

File: src/app/Home/book/show-books/show-books.component.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        alert(err.message)
```
Comment (ban-wtag) [inline src/app/Home/book/show-books/show-books.component.ts]: try to implement toaster service using [ngx-toastr](https://www.npmjs.com/package/ngx-toastr/v/16.0.1?activeTab=readme). Thank you

**Review by ban-wtag — APPROVED** [review]

**Review by hhu-wtag — APPROVED** [review]

**Pull request merged**